server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/GetSnapshotsResponse.java

```diff
@@ -171,59 +96,57 @@ public boolean isFailed() {
     }
 
     @Override
-    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
-        builder.startObject();
-        builder.startArray("responses");
-
-        for (Map.Entry<String, List<SnapshotInfo>> snapshots : successfulResponses.entrySet()) {
-            builder.startObject();
-            builder.field("repository", snapshots.getKey());
-            builder.startArray("snapshots");
-            for (SnapshotInfo snapshot : snapshots.getValue()) {
-                snapshot.toXContent(builder, params);
+    public void writeTo(StreamOutput out) throws IOException {
+        if (out.getVersion().onOrAfter(GetSnapshotsRequest.MULTIPLE_REPOSITORIES_SUPPORT_ADDED)) {
+            out.writeMap(failedResponses, StreamOutput::writeString, StreamOutput::writeException);
+        } else {
+            if (failedResponses.isEmpty() == false) {
+                assert false : "transport action should have thrown directly for old version but saw " + failedResponses;
+                throw failedResponses.values().iterator().next();
             }
-            builder.endArray();
-            builder.endObject();
         }
+        out.writeList(snapshots);
+    }
 
-        for (Map.Entry<String, ElasticsearchException> error : failedResponses.entrySet()) {
-            builder.startObject();
-            builder.field("repository", error.getKey());
-            ElasticsearchException.generateFailureXContent(builder, params, error.getValue(), true);
-            builder.endObject();
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, ToXContent.Params params) throws IOException {
+        builder.startObject();
+        builder.startArray("snapshots");
+        for (SnapshotInfo snapshotInfo : snapshots) {
+            snapshotInfo.toXContent(builder, params);
         }
-
         builder.endArray();
+        if (failedResponses.isEmpty() == false) {
+            builder.startObject("failures");
+            for (Map.Entry<String, ElasticsearchException> error : failedResponses.entrySet()) {
+                builder.field(error.getKey(), (b, pa) -> {
+                    b.startObject();
+                    error.getValue().toXContent(b, pa);
```

Comment:
Looks like we go from detailed (`ElasticsearchException.generateFailureXContent`) to some other form of exception output. I did not dig into the difference, but if you did, perhaps you can summarize it here?

Reply:
As far as I can see `generateFailureXContent` is more geared towards requests for multiple shards with it's `guessRootCauses` that would render a `root_cause` field which does not apply to our use case. Other than that it's the same format via `generateThrowableXContent` except that it's put at another level of nesting into an `error:` field which seems unnecessary for our use case of having a map by repo name.